### PR TITLE
FeatureVector.fromOriginal now sorts in place

### PR DESCRIPTION
**Pull request by fabuzaid21**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala [hunk 1]
```diff
@@ -966,4 +904,142 @@ private[ml] object AltDT extends Logging {
       PartitionInfo(newColumns, newNodeOffsets.flatten, newActiveNodes)
     }
   }
+
+  /**
+    * Feature vector types are based on (feature type, representation).
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala]: Update your IntelliJ settings:
Preferences -> Editor -> Code Style -> Scala - > ScalaDoc tab -> uncheck "Use scaladoc indent for leading asterisk"

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala [hunk 2]
```diff
@@ -126,6 +127,22 @@ class AltDTSuite extends SparkFunSuite with MLlibTestSparkContext  {
     assert(v === v2)
   }
 
+  test("FeatureVectorSortByValue") {
+    val values = Array(0.1, 0.2, 0.4, 0.6, 0.7, 0.9, 1.5, 1.55)
+    val col = Random.shuffle(values.toIterator).toArray
+    val (_, unsortedIndices) = col.zipWithIndex.unzip
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala]: Simplify: ```val unsortedIndices = col.indices```

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala [hunk 3]
```diff
@@ -126,6 +127,22 @@ class AltDTSuite extends SparkFunSuite with MLlibTestSparkContext  {
...
+    val sortedIndices = unsortedIndices.sortBy(x => col(x)).toArray
+    val fvUnsorted = Vectors.dense(col)
+    val featureIndex = 3
+    val featureArity = 5
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala]: arity 0

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala [hunk 4]
```diff
@@ -126,6 +127,22 @@ class AltDTSuite extends SparkFunSuite with MLlibTestSparkContext  {
...
+    val fvSorted =
+      FeatureVector.fromOriginal(featureIndex, featureArity, fvUnsorted)
+    assert(fvSorted.featureIndex == featureIndex)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala]: Use triple equals (better habit): ```===``` here and in lines below

**Comment by jkbradley** [comment]
To confirm: Is this what you ran the tests for the paper on?  And do you have a record of how much it improved the running time and/or memory usage?

**Comment by jkbradley** [comment]
I'll check out the other patch too

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala (hunk 1 above)
Comment (fabuzaid21) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala]: Hmmm…wasn't able to find a ScalaDoc tab – I'll just get rid of it manually

**Comment by fabuzaid21** [comment]
No, I ran the tests using DualPivotQuicksort, which I submitted in a pull request a long time ago. I'd have to re-run the experiments to see if the time saved using this matches the time saved before.

**Comment by jkbradley** [comment]
OK no problem.  I might run some quick tests to get some numbers.

**Comment by jkbradley** [comment]
I ran tests, and this was significantly faster for the sort (about 2x for 1M rows, 100 features, and about 1.5x for 1M rows, 1000 features).
Merging
Thanks!

**Pull request merged**